2020/day-5-binary-boarding/index.ts

```diff
@@ -0,0 +1,71 @@
+const firstHalf = (arr: any[]) => arr.slice(0, arr.length / 2)
+const lastHalf = (arr: any[]) => arr.slice(arr.length / 2)
+
+type State = {
+  row: number[]
+  col: number[]
+}
+
+type Char = 'F' | 'B' | 'L' | 'R'
+
+const reducer = (state: State, char: Char) => {
+  switch (char) {
+    case 'F': {
+      return {
+        ...state,
+        row: firstHalf(state.row),
+      }
+    }
+
+    case 'B': {
+      return {
+        ...state,
+        row: lastHalf(state.row),
+      }
+    }
+
+    case 'L': {
+      return {
+        ...state,
+        col: firstHalf(state.col),
+      }
+    }
+
+    case 'R': {
+      return {
+        ...state,
+        col: lastHalf(state.col),
+      }
+    }
+
+    default: {
+      throw new Error('Unknown code character')
+    }
+  }
+}
+
+export const findSeatPosition = (code: string) => {
+  const totalRows = Array.from(Array(128), (_, i) => i)
+  const totalCols = Array.from(Array(8), (_, i) => i)
+
+  let state = {
+    row: totalRows,
+    col: totalCols,
+  }
+
+  code.split('').forEach(char => {
+    state = reducer(state, char as Char)
```

Comment:
What is the `char as Char` doing here?

Reply:
This is a typecast - it overrides TypeScript's inferred type, which you usually don't want to do, but sometimes you have to and sometimes it's safe.

The reducer's 2nd parameter is typed as `Char` here:

```ts
type Char = 'F' | 'B' | 'L' | 'R'

const reducer = (state: State, char: Char) => {
```

But `char` here is inferred as a `string`, since `.split` returns a `string[]`, so without this casting, TS will throw this error: `Argument of type 'string' is not assignable to parameter of type 'Char'.`.

I think it's fine here since we know that the strings only contain these characters. We could perhaps make it slightly more robust by building a type predicate which checks the string to make sure there are no characters that aren't expected and then tells TS that it's definitely of type `Char`.. it crossed my mind but I felt like it was a bit much for AOC.

Actually as I'm writing this I'm thinking it's probably slightly more accurate to just remove the `Char` type and type the reducer parameter as a `string`, then the problem goes away